# <font color=blue>Assignments for "Logistic Regression"</font>

- In this assignment, you are going to use your model to predict survival of the Titanic disaster. For this assignment, send a link to a Jupyter notebook containing solutions to the following tasks.
    - Download [Titanic](https://www.kaggle.com/c/titanic/data) data from Kaggle. The data in the train.csv file meets your need.
    - Split your data into training and test sets.
    - Predict the survival based on the test data you split by creating your model.
    - Is your model's performance satisfactory? Explain.
    - Try to improve your model's performance by adding or subtracting some variables. <br>

- Explore the advantages and disadvantages of Logistic Regression and discuss with your mentor.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings(action="ignore")

In [2]:
titanic = pd.read_csv("C:/Users/Elif/data/titanic.csv")
titanic.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [3]:
titanic.isnull().sum()*100/titanic.shape[0]

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

In [4]:
titanic = titanic.drop(['Cabin'], axis=1)

In [5]:
titanic.dropna(inplace= True)

In [6]:
titanic.isnull().sum()*100/titanic.shape[0]

PassengerId    0.0
Survived       0.0
Pclass         0.0
Name           0.0
Sex            0.0
Age            0.0
SibSp          0.0
Parch          0.0
Ticket         0.0
Fare           0.0
Embarked       0.0
dtype: float64

In [7]:
titanic.corr()

PassengerId  Survived    Pclass       Age     SibSp     Parch  \
PassengerId     1.000000  0.029526 -0.035609  0.033681 -0.082704 -0.011672   
Survived        0.029526  1.000000 -0.356462 -0.082446 -0.015523  0.095265   
Pclass         -0.035609 -0.356462  1.000000 -0.365902  0.065187  0.023666   
Age             0.033681 -0.082446 -0.365902  1.000000 -0.307351 -0.187896   
SibSp          -0.082704 -0.015523  0.065187 -0.307351  1.000000  0.383338   
Parch          -0.011672  0.095265  0.023666 -0.187896  0.383338  1.000000   
Fare            0.009655  0.266100 -0.552893  0.093143  0.139860  0.206624   

                 Fare  
PassengerId  0.009655  
Survived     0.266100  
Pclass      -0.552893  
Age          0.093143  
SibSp        0.139860  
Parch        0.206624  
Fare         1.000000

In [8]:
titanic["is_female"] = pd.get_dummies(titanic.Sex).female
titanic["is_male"] = pd.get_dummies(titanic.Sex).male

In [9]:
X = titanic[["Pclass","SibSp","Age","Parch","is_male","Fare"]]

Y = titanic["Survived"]

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20, random_state=42)

In [11]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()
logreg.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [12]:
train_accuracy = logreg.score(X_train, y_train)
test_accuracy = logreg.score(X_test, y_test)
print('Accuracy of test data ',test_accuracy)

Accuracy of test data  0.7902097902097902


In [13]:
logreg_1 = LogisticRegression(multi_class="multinomial")
logreg_1.fit(X_train,y_train)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='multinomial', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [14]:
train_accuracy_softmax = logreg_1.score(X_train,y_train)
test_accuracy_softmax = logreg_1.score(X_test,y_test)
print("Softmax \n Train accuracy: {} \n Test accuracy: {}".format(train_accuracy_softmax,test_accuracy_softmax))

Softmax 
 Train accuracy: 0.8172231985940246 
 Test accuracy: 0.7972027972027972


Accuracy of my model 0.79 and it is not too bad but it must be improved.